### PR TITLE
[camera]remove CAS operation and use dispatch queue instead

### PR DESCRIPTION
**Pull request by hellohuanlin**
This PR removes the CompareAndSwap (CAS) operation and replaces it with a dispatch queue, as discussed in the [proposal](https://docs.google.com/document/d/1Znjp-BduXIb1Ix3JOE1YKyJi1Z0d6rB-2q6Lt7j8xSo/edit#). 

Also removed `volatile` keyword since dispatch queues are already memory barriers. 

I only added tests related to `latestPixelBuffer` which is the scope of this PR. But in the future we should add more sample buffer tests because it's one of the most complicated piece for this plugin, which it has almost zero coverage. 

This should be the last required item listed in https://github.com/flutter/flutter/issues/96429. 

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+@import camera;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface CameraTestUtils : NSObject
```
Comment (stuartmorgan) [inline packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h:9]: Please don't make fake classes that aren't actually classes just to hold what are effectively just functions; ObjC isn't Java, and has support for free-standing functions.

File: packages/camera/camera/example/ios/RunnerTests/FLTCamSampleBufferTests.m [hunk 2]
```diff
@@ -7,31 +7,35 @@
...
+- (void)testCopyPixelBuffer {
+  FLTCam *cam =
+      [CameraTestUtils createFLTCamWithCaptureSessionQueue:dispatch_queue_create("test", NULL)];
+  CMSampleBufferRef capturedSampleBuffer = [CameraTestUtils createTestSampleBuffer];
+  CVPixelBufferRef capturedPixelBuffer = CMSampleBufferGetImageBuffer(capturedSampleBuffer);
+  // Mimic sample buffer callback when captured a new video sample
+  [cam captureOutput:cam.captureVideoOutput
+      didOutputSampleBuffer:capturedSampleBuffer
+             fromConnection:OCMClassMock([AVCaptureConnection class])];
+  CVPixelBufferRef deliveriedPixelBuffer = [cam copyPixelBuffer];
```
Comment (stuartmorgan) [inline packages/camera/camera/example/ios/RunnerTests/FLTCamSampleBufferTests.m:35]: Don't you need to release this?

File: packages/camera/camera/ios/Classes/FLTCam.m [hunk 3]
```diff
@@ -76,6 +78,9 @@ @interface FLTCam () <AVCaptureVideoDataOutputSampleBufferDelegate,
 @property AVAssetWriterInputPixelBufferAdaptor *videoAdaptor;
 /// All FLTCam's state access and capture session related operations should be on run on this queue.
 @property(strong, nonatomic) dispatch_queue_t captureSessionQueue;
+/// The queue on which `latestPixelBuffer` property is accessed.
+/// To avoid unnecessary contention, we should not reuse the captureSessionQueue.
```
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/FLTCam.m:82]: What does "re-use" mean here?

(Also, go/avoidwe)

File: packages/camera/camera/ios/Classes/FLTCam.m [hunk 4]
```diff
@@ -355,10 +362,15 @@ - (void)captureOutput:(AVCaptureOutput *)output
   if (output == _captureVideoOutput) {
     CVPixelBufferRef newBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
     CFRetain(newBuffer);
-    CVPixelBufferRef old = _latestPixelBuffer;
-    while (!OSAtomicCompareAndSwapPtrBarrier(old, newBuffer, (void **)&_latestPixelBuffer)) {
-      old = _latestPixelBuffer;
-    }
+
+    __block CVPixelBufferRef old = nil;
```
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/FLTCam.m:366]: Please give this a more meaningful name while you're updating this code (e.g., `previousPixelBuffer`).

File: packages/camera/camera/ios/Classes/FLTCam_Test.h [hunk 5]
```diff
@@ -22,4 +22,10 @@
 @property(readonly, nonatomic)
     NSMutableDictionary<NSNumber *, FLTSavePhotoDelegate *> *inProgressSavePhotoDelegates;
 
+/// Delegate callback when receive a new video or audio sample.
```
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/FLTCam_Test.h:25]: Nit: receiving

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h (hunk 1 above)
Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h]: I changed to functions but kept the filename `CameraTestUtils`. There doesn't seem to be a rule about the file name for collection of functions in our [style guide](https://google.github.io/styleguide/objcguide.html#file-names). What do you think?

File: packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.m [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#import <OCMock/OCMock.h>
+@import AVFoundation;
+
+FLTCam *FLTCreateFLTCamWithCaptureSessionQueue(dispatch_queue_t captureSessionQueue) {
+  id inputMock = OCMClassMock([AVCaptureDeviceInput class]);
+  OCMStub([inputMock deviceInputWithDevice:[OCMArg any] error:[OCMArg setTo:nil]])
+      .andReturn(inputMock);
+
+  id sessionMock = OCMClassMock([AVCaptureSession class]);
+  OCMStub([sessionMock alloc]).andReturn(sessionMock);
```
Comment (stuartmorgan) [inline packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.m:15]: This is pretty ugly; please make an init that takes a capture session and expose that in the test method, then have the public init delegate to that one.

File: packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+// found in the LICENSE file.
+
+@import camera;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// Creates an `FLTCam` that runs its capture session operations on a given queue.
+/// @param captureSessionQueue the capture session queue
+/// @return an FLTCam object.
+extern FLTCam *FLTCreateFLTCamWithCaptureSessionQueue(dispatch_queue_t captureSessionQueue);
```
Comment (stuartmorgan) [inline packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h:12]: Optional nit: I don't think repeating FLT the second time is all that useful here; `FLTCreateCamWith...` seems clear in context.

(Modulo `Cam` not being a great name for a class in the first place.)

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.m (hunk 6 above)
Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.m:15]: I agree that it's better to inject these dependencies. As we write more unit tests, there are more dependencies to be injected. This can be a good starting point to adopt this pattern.

File: packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h (hunk 7 above)
Comment (hellohuanlin) [reply packages/camera/camera/example/ios/RunnerTests/CameraTestUtils.h:12]: Will remove `FLT`. 

I don't like the `Cam` name either, but will keep it for now in this PR.

File: packages/camera/camera/ios/Classes/FLTCam.h [hunk 8]
```diff
@@ -31,6 +31,13 @@ NS_ASSUME_NONNULL_BEGIN
 // Format used for video and image streaming.
 @property(assign, nonatomic) FourCharCode videoFormat;
 
+/// Init an `FLTCam` instance.
```
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/FLTCam.h:34]: Nit: Initializes

File: packages/camera/camera/ios/Classes/FLTCam.m [hunk 9]
```diff
@@ -92,6 +97,23 @@ - (instancetype)initWithCameraName:(NSString *)cameraName
                        orientation:(UIDeviceOrientation)orientation
                captureSessionQueue:(dispatch_queue_t)captureSessionQueue
                              error:(NSError **)error {
+  AVCaptureSession *captureSession = [[AVCaptureSession alloc] init];
```
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/FLTCam.m:100]: Optional nit: There's no reason to make a local variable for this; you can just inline the alloc/init in the method call.

File: packages/camera/camera/ios/Classes/FLTCam_Test.h [hunk 10]
```diff
@@ -17,9 +17,25 @@
...
 @property(readonly, nonatomic)
     NSMutableDictionary<NSNumber *, FLTSavePhotoDelegate *> *inProgressSavePhotoDelegates;
 
+/// Delegate callback when receiving a new video or audio sample.
+/// Exposed for unit tests.
+- (void)captureOutput:(AVCaptureOutput *)output
+    didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer
+           fromConnection:(AVCaptureConnection *)connection;
+
+/// Init function exposed for unit test.
```
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/FLTCam_Test.h:31]: "Initializes a camera instance. Allows for injecting dependencies that are usually internal."

File: packages/camera/camera/ios/Classes/FLTCam_Test.h [hunk 11]
```diff
@@ -17,9 +17,25 @@
...
+/// `captureSession` is injected to avoid mocking `alloc` function.
```
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/FLTCam_Test.h:32]: Please remove this; there's no reason to justify a common practice (allowing DI via initializers) in contrast to a rare practice (mocking alloc, which I don't think I've ever seen other than this plugin).

**Review by stuartmorgan — APPROVED** [review]
LGTM with nits.

**Review by jmagman — APPROVED** [review]
